helix-term/src/commands.rs

```diff
@@ -5367,14 +5369,40 @@ fn expand_selection(cx: &mut Context) {
 
         if let Some(syntax) = doc.syntax() {
             let text = doc.text().slice(..);
-            let selection = object::expand_selection(syntax, text, doc.selection(view.id));
+
+            let current_selection = doc.selection(view.id);
+
+            // save current selection so it can be restored using shrink_selection
+            view.object_selections.push(current_selection.clone());
+
+            let selection = object::expand_selection(syntax, text, current_selection);
             doc.set_selection(view.id, selection);
         }
     };
     motion(cx.editor);
     cx.editor.last_motion = Some(Motion(Box::new(motion)));
 }
 
+fn shrink_selection(cx: &mut Context) {
+    let motion = |editor: &mut Editor| {
+        let (view, doc) = current!(editor);
+        let selection = doc.selection(view.id);
+        // try to restore previous selection
+        if let Some(prev_selection) = view.object_selections.pop() {
+            // allow shrinking the selection only if current selection contains
+            // the previous object selection
+            if selection.contains(&prev_selection) {
+                doc.set_selection(view.id, prev_selection);
+            } else {
+                // clear existing selection as they can't be shrinked to anyway
+                view.object_selections.clear()
```

Comment:
Can you add that behavior in here? It should call `object::shrink_selection` and just look for the first `descendant_for_byte_range`

Reply:
Done, can you please check whether I understood the requirement correctly?

Reply:
It looks right to me now 🙂